File: packages/protocol/src/workspace.ts
/**
 * 工作区相关共享 DTO
 *
 * 用于 /api/workspaces/* 路由的请求与响应类型；前后端复用此定义。
 *
 * 设计要点：
 *   1. WorkspaceSpec 是当前活动工作区的事实源，包含 source、folders[]、settings 与不支持字段；
 *   2. 兼容 VSCode 风格的 .code-workspace 文件（仅 folders + settings 子集）；
 *   3. 路径安全：每个 WorkspaceFolder 解析后得到独立 rootId，所有文件读写必须绑定到某个 folderId；
 *   4. 不含历史兼容字段，命名直接采用最新方案。
 */

/**
 * 工作区来源类型。
 *   - directory       ：用户打开的是单个目录
 *   - code-workspace  ：用户打开的是 VSCode 兼容 .code-workspace 文件
 *   - fallback        ：无任何用户选择时使用的默认工作区（仓库内 ./workspace）
 */
export type WorkspaceSourceKind = 'directory' | 'code-workspace' | 'fallback';

/**
 * 工作区中的单个根文件夹。
 *
 * id 由后端生成；前端在所有文件 API 调用中通过 folderId 指明落点。
 */
export interface WorkspaceFolderSpec {
  /** 后端生成的稳定 ID，如 'wf-0'；同一会话内不变 */
  id: string;
  /** 展示名；优先取 .code-workspace 中的 name，否则取目录基名 */
  name: string;
  /** 解析后的绝对路径（POSIX 风格） */
  absolutePath: string;
  /** .code-workspace 中的原始 path 字段；directory 来源时与 absolutePath 一致 */
  originalPath: string;
  /** 原始 path 是否为绝对路径；用于 UI 风险提示 */
  isAbsolute: boolean;
}

/**
 * .code-workspace 中我们当前不解析、但保留作为兼容提示的字段名。
 * 例如 'extensions' / 'tasks' / 'launch'。
 */
export interface UnsupportedField {
  /** 字段名 */
  key: string;
  /** 该字段在 JSON 中的类型摘要，如 'object' / 'array' */
  kind: string;
}

/**
 * 当前活动工作区的完整描述。
 */
export interface WorkspaceSpec {
  /** 工作区 ID，按打开次数递增；fallback 工作区固定为 'fallback' */
  id: string;
  /** 工作区名；directory 取目录基名；code-workspace 取文件名（去 .code-workspace 后缀） */
  name: string;
  /** 工作区来源 */
  source: WorkspaceSourceKind;
  /** 当 source 为 'code-workspace' 时存在，是 .code-workspace 文件的绝对路径 */
  sourcePath: string | null;
  /** 解析后的 folders 列表；至少包含 1 个 */
  folders: WorkspaceFolderSpec[];
  /** DeepCode 命名空间下的工作区级设置（透传 .code-workspace.settings 中以 'deepcode.' 开头的键） */
  settings: Record<string, unknown>;
  /** 不支持但保留的字段，用于 UI 提示 */
  unsupportedFields: UnsupportedField[];
  /** 工作区打开时间（ISO） */
  openedAt: string;
}

/**
 * 工作区状态摘要，用于健康检查与启动日志。
 */
export interface WorkspaceSummary {
  id: string;
  name: string;
  source: WorkspaceSourceKind;
  /** folders 数量；首期常为 1 */
  folderCount: number;
}

/** GET /api/workspaces/current 成功响应 data 字段 */
export interface WorkspaceState {
  current: WorkspaceSpec;
  /** 当前是否使用 fallback 工作区（首次启动或上次工作区不可用） */
  fallbackUsed: boolean;
  /** 上一次 openWorkspace 失败原因；成功后清空 */
  lastError: string | null;
}

/** POST /api/workspaces/open 请求体 */
export interface OpenWorkspaceRequest {
  /**
   * 用户选择的绝对路径；可以指向：
   *   - 一个目录
   *   - 一个 .code-workspace 文件
   *
   * 必须是绝对路径；该路由是工作区入口，不受 /api/files/* 的相对路径约束。
   */
  path: string;
}

/** POST /api/workspaces/open 成功响应 data 字段 */
export interface OpenWorkspaceResult {
  workspace: WorkspaceSpec;
}

/** POST /api/workspaces/save-file 请求体 */
export interface SaveWorkspaceFileRequest {
  /** 目标 WorkspaceFolder；省略时使用当前 active/default folder */
  folderId?: string;
  /** 可选文件名；省略时使用当前工作区名生成 <name>.code-workspace */
  fileName?: string;
}

/** POST /api/workspaces/save-file 成功响应 data 字段 */
export interface SaveWorkspaceFileResult {
  /** 生成或覆盖的 .code-workspace 文件绝对路径 */
  workspaceFilePath: string;
  /** 保存后重新打开的工作区状态 */
  workspace: WorkspaceSpec;
  /** 本次是否创建了新文件 */
  created: boolean;
  /** 本次是否覆盖了已有文件 */
  overwritten: boolean;
}

/** PATCH /api/workspaces/current/settings 请求体 */
export interface PatchWorkspaceSettingsRequest {
  /**
   * 仅允许 DeepCode 命名空间下的键（前缀 'deepcode.'）；
   * 后端会拒绝其他键，避免误写入 VSCode 通用配置。
   */
  settings: Record<string, unknown>;
}

/** PATCH /api/workspaces/current/settings 成功响应 data 字段 */
export interface PatchWorkspaceSettingsResult {
  /** 合并后的最新 settings */
  settings: Record<string, unknown>;
}

// ============================================================================
// 文件系统浏览（用于"Open Workspace"可视化对话框）
// ============================================================================

/**
 * 浏览目录的单个子项。
 *
 * 该 DTO 仅用于 /api/fs/browse 路由：在用户点击 Open Workspace 时由前端
 * 模态对话框驱动进行目录浏览，最终由用户确认目录或 .code-workspace 文件，
 * 再走 POST /api/workspaces/open 完成切换。
 *
 * 注意：这里不暴露文件大小、修改时间等细节，避免成为通用文件系统 API；
 * 仅满足"打开工作区/打开 .code-workspace"场景所需的最小信息。
 */
export interface BrowseEntry {
  /** 条目名 */
  name: string;
  /** 解析后的绝对路径（POSIX 风格） */
  absolutePath: string;
  /** 条目类型 */
  type: 'directory' | 'file';
  /** 文件名是否以 .code-workspace 结尾；UI 据此突出显示 */
  isCodeWorkspace: boolean;
  /** 是否为隐藏项（以 . 开头），UI 默认折叠 */
  hidden: boolean;
}

/** GET /api/fs/browse 查询参数 */
export interface BrowsePathQuery {
  /**
   * 要浏览的绝对路径；不传或空字符串时由后端返回首选起点（用户主目录）。
   * 与 /api/files/* 不同，本路由出于"打开工作区"场景必须接受绝对路径。
   */
  path?: string;
}

/** GET /api/fs/browse 成功响应 data 字段 */
export interface BrowsePathResult {
  /** 实际解析到的绝对路径（POSIX 风格） */
  absolutePath: string;
  /** 父目录绝对路径；位于驱动器/文件系统根时为 null */
  parentPath: string | null;
  /** 子项列表（已按 directory 优先 + 名称升序排序） */
  entries: BrowseEntry[];
}

/**
 * 用户主目录、Windows 盘符、当前工作区父目录等"快捷起点"。
 * 由 GET /api/fs/initial-locations 返回，用于对话框侧栏。
 */
export interface InitialLocation {
  /** 展示标签，例如 "Home"、"E:\\"、"Current Workspace" */
  label: string;
  /** 绝对路径（POSIX 风格） */
  absolutePath: string;
  /** 分组：home / drive / workspace */
  kind: 'home' | 'drive' | 'workspace';
}

/** GET /api/fs/initial-locations 成功响应 data 字段 */
export interface InitialLocations {
  /**
   * 平台标识，与 Node.js `process.platform` 取值一致；
   * 这里用字符串字面量集合表达，避免 protocol 包依赖 @types/node。
   * 常见值：'win32' | 'linux' | 'darwin'，其余按字符串透传。
   */
  platform: 'win32' | 'linux' | 'darwin' | 'aix' | 'freebsd' | 'openbsd' | 'sunos' | 'cygwin' | 'netbsd' | string;
  /** 推荐起点列表（按重要性排序） */
  locations: InitialLocation[];
}

File: tauri/src-tauri/src/workspace.rs
// workspace.rs
//
// 工作区管理逻辑
//
// 维护当前活动工作区的 folders、name、source 等状态，
// 供 commands.rs 调用，避免 command 函数堆实现细节。

use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::sync::Mutex;

// ---- DTO：与 protocol DTO 字段同构 ----

/// 工作区来源类型
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum WorkspaceSourceKind {
    Directory,
    CodeWorkspace,
    Fallback,
}

/// 工作区中的单个根文件夹
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceFolderSpec {
    pub id: String,
    pub name: String,
    pub absolute_path: String,
    pub original_path: String,
    pub is_absolute: bool,
}

/// 不支持但保留的字段
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnsupportedField {
    pub key: String,
    pub kind: String,
}

/// 当前活动工作区完整描述
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceSpec {
    pub id: String,
    pub name: String,
    pub source: WorkspaceSourceKind,
    pub source_path: Option<String>,
    pub folders: Vec<WorkspaceFolderSpec>,
    pub settings: serde_json::Value,
    pub unsupported_fields: Vec<UnsupportedField>,
    pub opened_at: String,
}

/// GET /api/workspaces/current 成功响应 data
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceState {
    pub current: WorkspaceSpec,
    pub fallback_used: bool,
    pub last_error: Option<String>,
}

/// POST /api/workspaces/open 成功响应 data
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenWorkspaceResult {
    pub workspace: WorkspaceSpec,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveWorkspaceFileResult {
    pub workspace_file_path: String,
    pub workspace: WorkspaceSpec,
    pub created: bool,
    pub overwritten: bool,
}

/// PATCH /api/workspaces/current/settings 成功响应 data
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatchWorkspaceSettingsResult {
    pub settings: serde_json::Value,
}

// ---- 工作区管理器 ----

/// 工作区管理器（Tauri managed state）
pub struct WorkspaceManager {
    inner: Mutex<WorkspaceManagerInner>,
}

struct WorkspaceManagerInner {
    /// 当前活动工作区；首次 open_workspace 之前为 None
    current: Option<WorkspaceSpec>,
    /// 打开计数器，用于生成 id
    next_id: u32,
}

impl WorkspaceManager {
    pub fn new() -> Self {
        Self {
            inner: Mutex::new(WorkspaceManagerInner {
                current: None,
                next_id: 1,
            }),
        }
    }

    /// 获取当前工作区状态；若未打开则初始化 fallback，保证后续文件操作能解析 wf-0
    pub fn get_current(&self) -> WorkspaceState {
        let mut inner = self.inner.lock().unwrap();
        if let Some(ws) = &inner.current {
            return WorkspaceState {
                current: ws.clone(),
                fallback_used: matches!(ws.source, WorkspaceSourceKind::Fallback),
                last_error: None,
            };
        }

        // 返回 fallback 工作区：当前工作目录，同时写入 current 作为事实源
        let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
        let folder_name = cwd
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_else(|| "workspace".into());
        let abs = to_posix(&cwd);
        let fallback = WorkspaceSpec {
            id: "fallback".into(),
            name: folder_name,
            source: WorkspaceSourceKind::Fallback,
            source_path: None,
            folders: vec![WorkspaceFolderSpec {
                id: "wf-0".into(),
                name: abs.split('/').last().unwrap_or("workspace").into(),
                absolute_path: abs.clone(),
                original_path: abs,
                is_absolute: true,
            }],
            settings: serde_json::json!({}),
            unsupported_fields: vec![],
            opened_at: chrono_now_iso(),
        };
        inner.current = Some(fallback.clone());
        WorkspaceState {
            current: fallback,
            fallback_used: true,
            last_error: None,
        }
    }

    /// 打开工作区（目录或 .code-workspace 文件）
    pub fn open_workspace(&self, path: &str) -> Result<OpenWorkspaceResult, String> {
        let p = PathBuf::from(path);
        if !p.exists() {
            return Err(format!("路径不存在: {}", path));
        }

        let abs = p
            .canonicalize()
            .map_err(|e| format!("路径规范化失败: {}", e))?;
        let abs_posix = to_posix(&abs);

        let (source, name, folders, source_path, settings, unsupported) =
            if abs.is_dir() {
                // 目录模式
                let dir_name = abs
                    .file_name()
                    .map(|n| n.to_string_lossy().to_string())
                    .unwrap_or_else(|| "workspace".into());
                (
                    WorkspaceSourceKind::Directory,
                    dir_name,
                    vec![WorkspaceFolderSpec {
                        id: "wf-0".into(),
                        name: abs
                            .file_name()
                            .map(|n| n.to_string_lossy().to_string())
                            .unwrap_or_else(|| "workspace".into()),
                        absolute_path: abs_posix.clone(),
                        original_path: abs_posix,
                        is_absolute: true,
                    }],
                    None::<String>,
                    serde_json::json!({}),
                    vec![] as Vec<UnsupportedField>,
                )
            } else if is_code_workspace_file(&abs) {
                // .code-workspace 文件模式
                let content = std::fs::read_to_string(&abs)
                    .map_err(|e| format!("读取 .code-workspace 失败: {}", e))?;
                let ws_json: serde_json::Value = serde_json::from_str(&content)
                    .map_err(|e| format!("解析 .code-workspace JSON 失败: {}", e))?;

                let file_name = abs
                    .file_stem()
                    .map(|n| n.to_string_lossy().to_string())
                    .unwrap_or_else(|| "workspace".into());

                let folders_val = ws_json.get("folders").and_then(|v| v.as_array());
                let parent = abs.parent().unwrap_or(Path::new("."));

                let mut folder_specs = Vec::new();
                if let Some(arr) = folders_val {
                    for (i, item) in arr.iter().enumerate() {
                        let raw_path = item
                            .get("path")
                            .and_then(|v| v.as_str())
                            .unwrap_or(".");
                        let resolved = if Path::new(raw_path).is_absolute() {
                            PathBuf::from(raw_path)
                        } else {
                            parent.join(raw_path)
                        };
                        let canonical = match resolved.canonicalize() {
                            Ok(c) => c,
                            Err(_) => resolved.clone(),
                        };
                        let folder_name = item
                            .get("name")
                            .and_then(|v| v.as_str())
                            .map(String::from)
                            .unwrap_or_else(|| {
                                canonical
                                    .file_name()
                                    .map(|n| n.to_string_lossy().to_string())
                                    .unwrap_or_else(|| format!("folder-{}", i))
                            });
                        folder_specs.push(WorkspaceFolderSpec {
                            id: format!("wf-{}", i),
                            name: folder_name,
                            absolute_path: to_posix(&canonical),
                            original_path: raw_path.into(),
                            is_absolute: Path::new(raw_path).is_absolute(),
                        });
                    }
                }

                if folder_specs.is_empty() {
                    folder_specs.push(WorkspaceFolderSpec {
                        id: "wf-0".into(),
                        name: file_name.clone(),
                        absolute_path: to_posix(parent),
                        original_path: ".".into(),
                        is_absolute: false,
                    });
                }

                // 提取 deepcode. 前缀的 settings
                let settings_obj = ws_json
                    .get("settings")
                    .and_then(|v| v.as_object())
                    .map(|obj| {
                        let mut dc = serde_json::Map::new();
                        for (k, v) in obj {
                            if k.starts_with("deepcode.") {
                                dc.insert(k.clone(), v.clone());
                            }
                        }
                        serde_json::Value::Object(dc)
                    })
                    .unwrap_or(serde_json::json!({}));

                // 识别不支持的字段
                let known_keys = ["folders", "settings"];
                let unsupported: Vec<UnsupportedField> = ws_json
                    .as_object()
                    .map(|obj| {
                        obj.iter()
                            .filter(|(k, _)| !known_keys.contains(&k.as_str()))
                            .map(|(k, v)| UnsupportedField {
                                key: k.clone(),
                                kind: match v {
                                    serde_json::Value::Object(_) => "object".into(),
                                    serde_json::Value::Array(_) => "array".into(),
                                    serde_json::Value::String(_) => "string".into(),
                                    serde_json::Value::Number(_) => "number".into(),
                                    serde_json::Value::Bool(_) => "boolean".into(),
                                    serde_json::Value::Null => "null".into(),
                                },
                            })
                            .collect()
                    })
                    .unwrap_or_default();

                (
                    WorkspaceSourceKind::CodeWorkspace,
                    file_name,
                    folder_specs,
                    Some(abs_posix),
                    settings_obj,
                    unsupported,
                )
            } else {
                return Err(format!(
                    "路径既不是目录也不是 .code-workspace 文件: {}",
                    path
                ));
            };

        let mut inner = self.inner.lock().unwrap();
        let id = format!("ws-{}", inner.next_id);
        inner.next_id += 1;

        let ws = WorkspaceSpec {
            id,
            name,
            source,
            source_path,
            folders,
            settings,
            unsupported_fields: unsupported,
            opened_at: chrono_now_iso(),
        };

        inner.current = Some(ws.clone());

        Ok(OpenWorkspaceResult { workspace: ws })
    }

    pub fn save_workspace_file(
        &self,
        folder_id: Option<String>,
        file_name: Option<String>,
    ) -> Result<SaveWorkspaceFileResult, String> {
        let (current, target_folder) = {
            let inner = self.inner.lock().unwrap();
            let ws = inner
                .current
                .as_ref()
                .ok_or_else(|| "工作区尚未初始化，请先调用 get_current_workspace".to_string())?
                .clone();
            let folder = match folder_id.as_deref() {
                Some(id) => ws
                    .folders
                    .iter()
                    .find(|f| f.id == id)
                    .cloned()
                    .ok_or_else(|| format!("未知 folderId: {}", id))?,
                None => ws
                    .folders
                    .first()
                    .cloned()
                    .ok_or_else(|| "当前工作区没有 folder".to_string())?,
            };
            (ws, folder)
        };

        let root = PathBuf::from(&target_folder.absolute_path);
        if !root.is_dir() {
            return Err(format!(
                "目标 folder 不存在或不是目录: {}",
                target_folder.absolute_path
            ));
        }

        let workspace_file_name =
            sanitize_workspace_file_name(file_name.as_deref().unwrap_or(&current.name));
        let workspace_file = root.join(workspace_file_name);
        if workspace_file.parent() != Some(root.as_path()) {
            return Err("workspace 文件只能保存到当前 folder 根目录".into());
        }

        let folders: Vec<serde_json::Value> = current
            .folders
            .iter()
            .map(|folder| {
                let path = workspace_folder_path_for_file(&root, folder, &target_folder.id);
                serde_json::json!({
                    "path": path,
                    "name": folder.name,
                })
            })
            .collect();
        let payload = serde_json::json!({
            "folders": folders,
            "settings": current.settings,
        });

        let existed = workspace_file.exists();
        let content = serde_json::to_string_pretty(&payload)
            .map_err(|e| format!("序列化 workspace 文件失败: {}", e))?;
        std::fs::write(&workspace_file, format!("{}\n", content))
            .map_err(|e| format!("写入 workspace 文件失败: {}", e))?;

        let workspace_file_path = to_posix(&workspace_file);
        let opened = self.open_workspace(&workspace_file_path)?;
        Ok(SaveWorkspaceFileResult {
            workspace_file_path,
            workspace: opened.workspace,
            created: !existed,
            overwritten: existed,
        })
    }

    /// 合并 DeepCode 命名空间设置（Tauri 内存态，与 Node server 行为对齐）
    pub fn patch_workspace_settings(
        &self,
        settings: serde_json::Value,
    ) -> Result<PatchWorkspaceSettingsResult, String> {
        let mut inner = self.inner.lock().unwrap();
        let ws = inner
            .current
            .as_mut()
            .ok_or_else(|| "工作区尚未初始化，请先调用 get_current_workspace".to_string())?;
        let patch_obj = settings
            .as_object()
            .ok_or_else(|| "settings 必须是对象".to_string())?;

        let mut next = match ws.settings.as_object() {
            Some(obj) => obj.clone(),
            None => serde_json::Map::new(),
        };
        let mut rejected: Vec<String> = Vec::new();
        for (key, value) in patch_obj {
            if !key.starts_with("deepcode.") {
                rejected.push(key.clone());
                continue;
            }
            next.insert(key.clone(), value.clone());
        }
        if !rejected.is_empty() {
            return Err(format!(
                "仅允许 'deepcode.' 前缀的设置键；被拒绝: {}",
                rejected.join(", ")
            ));
        }

        ws.settings = serde_json::Value::Object(next.clone());
        Ok(PatchWorkspaceSettingsResult {
            settings: serde_json::Value::Object(next),
        })
    }

    /// 按 folderId 获取 folder 的绝对路径
    pub fn get_folder_abs_path(&self, folder_id: &str) -> Option<String> {
        let inner = self.inner.lock().unwrap();
        inner.current.as_ref().and_then(|ws| {
            ws.folders
                .iter()
                .find(|f| f.id == folder_id)
                .map(|f| f.absolute_path.clone())
        })
    }
}

// ---- helper ----

/// 路径转 POSIX 风格字符串
fn to_posix(p: &Path) -> String {
    p.to_string_lossy().replace('\\', "/")
}

/// 判断是否为 .code-workspace 文件
fn is_code_workspace_file(p: &Path) -> bool {
    p.file_name()
        .map(|name| {
            name.to_string_lossy()
                .to_lowercase()
                .ends_with(".code-workspace")
        })
        .unwrap_or(false)
}

fn sanitize_workspace_file_name(input: &str) -> String {
    let trimmed = input.trim();
    let raw = if trimmed.is_empty() {
        "workspace.code-workspace".to_string()
    } else if trimmed.to_lowercase().ends_with(".code-workspace") {
        trimmed.to_string()
    } else {
        format!("{}.code-workspace", trimmed)
    };
    let sanitized: String = raw
        .chars()
        .map(|ch| match ch {
            '<' | '>' | ':' | '"' | '/' | '\\' | '|' | '?' | '*' => '-',
            ch if ch.is_control() => '-',
            ch => ch,
        })
        .collect();
    if sanitized == ".code-workspace" {
        "workspace.code-workspace".into()
    } else {
        sanitized
    }
}

fn workspace_folder_path_for_file(
    workspace_file_dir: &Path,
    folder: &WorkspaceFolderSpec,
    target_folder_id: &str,
) -> String {
    if folder.id == target_folder_id {
        return ".".into();
    }
    let folder_abs = PathBuf::from(&folder.absolute_path);
    match folder_abs.strip_prefix(workspace_file_dir) {
        Ok(relative) if relative.as_os_str().is_empty() => ".".into(),
        Ok(relative) => to_posix(relative),
        Err(_) => folder.absolute_path.clone(),
    }
}

/// 当前时间 ISO 8601 格式（UTC）
fn chrono_now_iso() -> String {
    chrono::Utc::now().to_rfc3339()
}
